app/src/main/cpp/skyline/kernel/svc.cpp

```diff
@@ -11,209 +11,250 @@
-        if (!object)
-            throw exception("svcMapMemory: Cannot find memory object in handle table for address 0x{:X}", source);
-        object->item->UpdatePermission(span<u8>{source, size}, {false, false, false});
+        state.process->memory.SetChunkPermission(span<u8>{source, size}, {false, false, false});
+        state.process->memory.SetLockOnChunks(span<u8>{source, size}, true);
 
         Logger::Debug("Mapped range 0x{:X} - 0x{:X} to 0x{:X} - 0x{:X} (Size: 0x{:X} bytes)", source, source + size, destination, destination + size, size);
         state.ctx->gpr.w0 = Result{};
     }
 
     void UnmapMemory(const DeviceState &state) {
-        auto source{reinterpret_cast<u8 *>(state.ctx->gpr.x0)};
-        auto destination{reinterpret_cast<u8 *>(state.ctx->gpr.x1)};
+        u8 *destination{reinterpret_cast<u8 *>(state.ctx->gpr.x0)};
+        u8 *source{reinterpret_cast<u8 *>(state.ctx->gpr.x1)};
         size_t size{state.ctx->gpr.x2};
 
-        if (!util::IsPageAligned(destination) || !util::IsPageAligned(source)) {
+        if (!util::IsPageAligned(destination) || !util::IsPageAligned(source)) [[unlikely]] {
             state.ctx->gpr.w0 = result::InvalidAddress;
-            Logger::Warn("Addresses not page aligned: Source: 0x{:X}, Destination: 0x{:X} (Size: 0x{:X} bytes)", source, destination, size);
+            Logger::Warn("Addresses not page aligned: 'source': 0x{:X}, 'destination': 0x{:X}, 'size': {} bytes", source, destination, size);
             return;
         }
 
-        if (!util::IsPageAligned(size)) {
+        if (!size || !util::IsPageAligned(size)) [[unlikely]] {
             state.ctx->gpr.w0 = result::InvalidSize;
-            Logger::Warn("'size' {}: 0x{:X}", size ? "not page aligned" : "is zero", size);
+            Logger::Warn("'size' {}: 0x{:X}", size ? "is not page aligned" : "is zero", size);
             return;
         }
 
-        auto stack{state.process->memory.stack};
-        if (!stack.contains(span<u8>{source, size})) {
+        if (!state.process->memory.stack.contains(span<u8>{destination, size})) [[unlikely]] {
             state.ctx->gpr.w0 = result::InvalidMemoryRegion;
-            Logger::Warn("Source not within stack region: Source: 0x{:X}, Destination: 0x{:X} (Size: 0x{:X} bytes)", source, destination, size);
+            Logger::Warn("Source not within stack region: 'source': 0x{:X}, 'destination': 0x{:X}, 'size': 0x{:X} bytes", source, destination, size);
             return;
         }
 
-        auto sourceChunk{state.process->memory.Get(source)};
-        auto destChunk{state.process->memory.Get(destination)};
-        if (!sourceChunk || !destChunk) {
-            state.ctx->gpr.w0 = result::InvalidAddress;
-            Logger::Warn("Addresses have no descriptor: Source: 0x{:X}, Destination: 0x{:X} (Size: 0x{:X} bytes)", source, destination, size);
-            return;
-        }
+        auto dstChunk{state.process->memory.GetChunk(destination).value()};
```

Comment:
This should surely be implemented in the memory manager, it seems racy otherwise